### DISCUSSION TITLE
Eloquent Validation Rule

### DISCUSSION THREAD

**Discussion by tam5** [laravel/framework#39235] [Ideas]
## Idea

There are a few validation rules that allow the use of the query builder, like `exists` and `unique`. However, they only provide a database query builder, rather than an eloquent query builder.

It would be great to be able to have the eloquent query builder to be able to make use of relationships when performing validation.

## Examples

A simple example taken from the docs:

```php
'email' => Rule::unique('users')->where(function ($query) {
    return $query->where('account_id', 1);
})
```

This works fine, but it feels weird if in other places of the codebase I would be doing `User::whereRelation('account', 'id', 1)`. Additionally, it's nice to be able to use the abstraction of the relationship, rather than relying on the underlying implemented foreign key column.

Instead it would be great if I could do something like:

```php
'email' => Rule::unique(User::class)->where(function ($query) {
    $query->whereRelation('account', 'id', 1);
})

// or 

'email' => Rule::unique(User::class)->whereRelation('account', 'id', 1);
```

Being able to access relationships also opens the door for more complex validation, like validating belongsTo relationships.

For example, suppose I have Users, Accounts, and User Groups. I should be able to add a user to a user group that belongs to the same account as the user, but I should not be able to add a user to a user group if the user is not part of that account.

Currently, that would require a good deal of manual work to set up a validation rule for that. However, if we were to have access to the eloquent builder, this becomes very easy.

Example for adding users to user group:

```php
'user_id' => Rule::exists(User::class)->whereRelation('accounts', 'id', $this->user_group->account_id);
```

## Extensions

Possible ways this could be expanded further, would be to have explicit rules that help validate relationships, which could make things nice and readable:

Examples:

```php
'belongs_to' => 'relation,column,value',
Rule::belongsTo($relation, $column, $operator, $value)

'has' => 'relation,column,value',
Rule::has($relation, $column, $operator, $value)
```

## Considerations

In my examples I've used the `unique` and `exists` rules directly for clarity, but it probably doesn't make much sense to return something different based on whether a model was passed or a table name. Instead we'd probably need to come up with different names for the rules that return an eloquent builder.

---

Anyway, looking for feedback here. If it seems useful I can start a PR.

**Comment by KaliaJS** [comment]
This is exactly what I'm looking for. What is the alternative currently ?

**Comment by shaedrich** [comment]
Maybe, this could even be used like `User::factory()` as `User::validationRule()->whereRelation('accounts', 'id', $this->user_group->account_id);`. This would have the benefit that `Rule::unique` would stay unchanged and didn't need to branch of a somewhat different functionality. Instead, `User::validationRule()` could either just set a flag and return `$this`, or it could return a higher order message of some sort, similar to those of collections. However, if existing projects have a method named "validationRule()", they would break with this change.
